libs/geometry/src/Transcoder.cpp

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (C) 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <geometry/Transcoder.h>
+
+#include <math/half.h>
+
+using filament::math::half;
+
+namespace filament {
+namespace geometry {
+
+// The internal workhorse function of the Transcoder, which takes arbitrary input but always
+// produced packed floats. We expose a more readable interface than this to users, who often have
+// untyped blobs of interleaved data. Note that this variant takes an arbitrary number of
+// components, we also have a fixed-size variant for better compiler output.
+template<typename SOURCE_TYPE, int NORMALIZATION_FACTOR>
+void convert(float* target, void const* source, size_t count, int numComponents,
```

Comment:
`UTILS_RESTRICT` pointers will help with vectorizing / unrolling.

Reply:
I *think* C++ 17 allows `float` template parameters (e.g. for NORMALIZATION_FACTOR)

Reply:
`numComponents` -> `componentCount`

Reply:
> I _think_ C++ 17 allows `float` template parameters (e.g. for NORMALIZATION_FACTOR)

I tried this and failed.

Reply:
Done